File: tools/gen_modelbin_web/events.py
"""Push events from Python straight into the page.

Replaces the Tkinter shell's `_poll_queue` (a queue.Queue polled every
100ms via root.after) entirely -- pywebview's window.evaluate_js is
callable from any thread, so a worker thread can push an event the
moment it has one instead of a poller picking it up later. The frontend
registers per-tag listeners on window.__forzaEvents (see
frontend/js/events.js); each listener does its own `generation`
staleness check, the same role the queue's (tag, generation, ...) tuple
comparison plays today (e.g. tabs/glyph_inspector.py's stale-result guard).
"""
from __future__ import annotations

import json
import sys
import time

# pywebview's WinForms/WebView2 backend has a startup race: calling
# evaluate_js from a background thread before the CoreWebView2Controller
# has fully finished initializing throws a cascade of COM-interop errors
# ("CoreWebView2Controller members can only be accessed from the UI
# thread", occasionally alongside a pythonnet recursion error walking
# window.native.AccessibilityObject). It's transient, not a real deadlock
# -- confirmed empirically it recovers on its own -- but a single
# fire-and-forget evaluate_js call has no way to know that, so it retries
# with backoff instead of silently losing the event.
_MAX_ATTEMPTS = 6
_INITIAL_BACKOFF = 0.3


def push_event(window, tag: str, generation: int, payload: dict) -> None:
    js = (
        "window.__forzaEvents && window.__forzaEvents.dispatch("
        f"{json.dumps(tag)}, {generation}, {json.dumps(payload)})"
    )
    delay = _INITIAL_BACKOFF
    for attempt in range(_MAX_ATTEMPTS):
        try:
            window.evaluate_js(js)
            return
        except Exception as exc:  # the WebView2 startup race described above
            if attempt == _MAX_ATTEMPTS - 1:
                print(f"push_event: giving up on {tag!r} after {_MAX_ATTEMPTS} attempts: {exc}",
                      file=sys.stderr)
                return
            time.sleep(delay)
            delay *= 2

File: tools/gen_modelbin_web/frontend/js/events.js
// Receiving end of events.py's push_event: a tiny per-tag pub-sub that
// Python pushes into directly via window.evaluate_js, replacing the old
// Tkinter shell's 100ms _poll_queue tick entirely. Each listener does its
// own `generation` staleness check -- the same role the old
// (tag, generation, ...) tuple comparison played in e.g.
// tabs/glyph_inspector.py's stale-result guard.
window.__forzaEvents = (function () {
  const listeners = {};
  return {
    on(tag, fn) {
      (listeners[tag] || (listeners[tag] = [])).push(fn);
      return () => {
        listeners[tag] = (listeners[tag] || []).filter((f) => f !== fn);
      };
    },
    dispatch(tag, generation, payload) {
      (listeners[tag] || []).forEach((fn) => {
        try { fn(generation, payload); } catch (e) { console.error(e); }
      });
    },
  };
})();
